src/docfx/build/toc/parser/MarkdownTocMarkup.cs

```diff
@@ -45,141 +43,111 @@ public static (List<Error> errors, TableOfContentsModel model) LoadMdTocModel(st
             var (validationErrors, tocMetadata) = JsonUtility.ToObject<TableOfContentsMetadata>(metadata);
             errors.AddRange(validationErrors);
 
-            var tocModel = new TableOfContentsModel { Metadata = tocMetadata };
-            tocModel.Items = ConvertTo(tocContent, file.FilePath, headingBlocks.ToArray(), errors).children;
+            var tree = BuildTree(errors, file.FilePath, headingBlocks);
+
+            var tocModel = new TableOfContentsModel { Metadata = tocMetadata, Items = tree.Items };
 
             return (errors, tocModel);
         }
 
-        private static (List<TableOfContentsItem> children, int count) ConvertTo(string tocContent, string filePath, HeadingBlock[] headingBlocks, List<Error> errors, int startIndex = 0)
+        private static TableOfContentsItem BuildTree(List<Error> errors, string filePath, List<HeadingBlock> blocks)
         {
-            if (headingBlocks.Length == 0)
-            {
-                return (new List<TableOfContentsItem>(), 0);
-            }
+            var result = new TableOfContentsItem();
+            var stack = new Stack<(int level, TableOfContentsItem item)>();
+            var top = (level: 0, node: result);
+            stack.Push(top);
 
-            Debug.Assert(startIndex < headingBlocks.Length);
-
-            int i = startIndex;
-            var items = new List<TableOfContentsItem>();
-            var childrenCount = 0;
-            do
+            foreach (var block in blocks)
             {
-                var item = GetItem(headingBlocks[i]);
-                if (item == null)
+                var currentLevel = block.Level;
+                var currentItem = GetItem(errors, filePath, block);
+                if (currentItem == null)
                 {
                     continue;
                 }
 
-                items.Add(item);
-                var currentLevel = headingBlocks[i].Level;
-                var (nextLevelDistance, skipped) = GetNextLevelDistance(i, headingBlocks[i].Level);
-                if (nextLevelDistance > 0)
+                while (stack.TryPeek(out top) && top.level >= currentLevel)
```

Comment:
>                 [](start = 0, length = 16)

Try some cases like:
##### 1
#### 2
### 3

Seems the skip level error will not be reported